ci/python-publish.yml

```
@@ -12,28 +12,42 @@ on:
   release:
     types: [published]
 
-permissions:
-  contents: read
-
 jobs:
-  deploy:
+  release-build:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v4
 
+      - uses: actions/setup-python@v5
+        with:
+          python-version: "3.x"
+
+      - name: build release distributions
+        run: |
+          # NOTE: put your own distribution build steps here.
+          python -m pip install build
+          python -m build
+
+      - name: upload distributions
+        uses: actions/upload-artifact@v4
+        with:
+          name: release-dists
+          path: dist/
+
+  pypi-publish:
     runs-on: ubuntu-latest
+    needs:
+      - release-build
+    permissions:
```

It'd *really* want to get people to use environment-based protection from the very beginning:
```suggestion
    environment:
      name: pypi
      url: https://pypi.org/p/<project-name>
    permissions:
```

---

Agreed, although this will produce an incongruity between this workflow and the other GitHub docs. I suppose we should update all of them to suggest using an environment.

---

All the docs and examples I could reach already use environments called `pypi` and `testpypi` by default. Not `release` since that name is ambiguous and doesn't represent a specific deployment target (the PyPI).

---

GH docs you contributed earlier seem to use the `release` environment (so does the placeholder in Warehouse, which I dislike for the same reason).

---

I was thinking of [this example](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-pypi#updating-your-github-actions-workflow), which (regrettably) doesn't use an environment. I'll add an environment here and work on updating that doc separately.

---

Added an environment example with some comments, PTAL!

---

Looks good. Normally I like making versioned URLs. But that might be confusing to newbies. Although, if there's an easily retrieveable version in the release event, sticking it into the example comment would be a nice touch.

---

I think there is, although there's no guarantee that the release version is a match for the PyPI version (user error, but also because of PEP 440 normalization/rewrites). So I'm inclined to leave this unversioned for now, although I don't have a super strong opinion here 🙂

---

I think it's fine if it's not normalized and the URL might be off (although, personally I started embedding `workflow_dispatch` inputs validation for PEP 440), and if it's in a comment, as a hint. The most reliable way of getting that would be unpacking a wheel or sdist and grabbing it from the structured metadata since the packaging machinery usually normalizes it already. I won't insist, though.